katharsis-jpa/src/main/java/io/katharsis/jpa/internal/query/backend/querydsl/QuerydslQueryBackend.java

```diff
@@ -1,386 +1,401 @@
-		if (expression instanceof CollectionExpressionBase) {
-			return ((CollectionExpressionBase) expression).getElementType();
-		}
-		return expression.getType();
-	}
-
-	@Override
-	public Expression<?> getAttribute(final Expression<?> expression, MetaAttribute pathElement) {
-		if (pathElement instanceof MetaVirtualAttribute) {
-			VirtualAttributeRegistry virtualAttrs = queryImpl.getVirtualAttrs();
-			QuerydslExpressionFactory expressionFactory = (QuerydslExpressionFactory) virtualAttrs
-					.get((MetaVirtualAttribute) pathElement);
-			return expressionFactory.getExpression(expression, getQuery());
-		}
-		else {
-			return QuerydslUtils.get(expression, pathElement.getName());
-		}
-	}
-
-	@Override
-	public Expression<?> joinSubType(Expression<?> expression, Class<?> entityClass) {
-		BeanPath beanPath = (BeanPath) expression;
-		Class<?> queryClass = QuerydslUtils.getQueryClass(entityClass);
-		return beanPath.as(queryClass);
-	}
-
-	@Override
-	public Expression<?> doJoin(MetaAttribute targetAttr, JoinType joinType, Expression<?> parent) {
-		if (targetAttr instanceof MetaVirtualAttribute) {
-
-			MetaVirtualAttribute virtualAttr = (MetaVirtualAttribute) targetAttr;
-			QuerydslExpressionFactory expressionFactory = (QuerydslExpressionFactory<?>) queryImpl.getVirtualAttrs()
-					.get(virtualAttr);
-
-			return expressionFactory.getExpression(parent, getQuery());
-		}
-		else {
-			Expression<Object> expression = QuerydslUtils.get(parent, targetAttr.getName());
-			querydslQuery.getMetadata().addJoin(QuerydslUtils.convertJoinType(joinType), expression);
-			return expression;
-		}
-	}
+package io.katharsis.jpa.internal.query.backend.querydsl;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import javax.annotation.Nullable;
+import javax.persistence.criteria.JoinType;
+
+import com.google.common.collect.ImmutableList;
+import com.querydsl.core.support.FetchableSubQueryBase;
+import com.querydsl.core.types.CollectionExpression;
+import com.querydsl.core.types.EntityPath;
+import com.querydsl.core.types.Expression;
+import com.querydsl.core.types.ExpressionUtils;
+import com.querydsl.core.types.OperationImpl;
+import com.querydsl.core.types.Ops;
+import com.querydsl.core.types.Order;
+import com.querydsl.core.types.OrderSpecifier;
+import com.querydsl.core.types.Path;
+import com.querydsl.core.types.Predicate;
+import com.querydsl.core.types.QTuple;
+import com.querydsl.core.types.dsl.BeanPath;
+import com.querydsl.core.types.dsl.CollectionExpressionBase;
+import com.querydsl.core.types.dsl.ComparableExpression;
+import com.querydsl.core.types.dsl.LiteralExpression;
+import com.querydsl.core.types.dsl.MapExpressionBase;
+import com.querydsl.core.types.dsl.MapPath;
+import com.querydsl.core.types.dsl.NumberExpression;
+import com.querydsl.core.types.dsl.SimpleExpression;
+import com.querydsl.core.types.dsl.StringExpression;
+import com.querydsl.jpa.JPAQueryBase;

```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'com.querydsl.jpa.JPAQueryBase'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AUselessImportCheck)